`PLW1514` does not check for uses of `pathlib.Path.open()` without an explicit encoding argument.

# Description

[PLW1514](https://docs.astral.sh/ruff/rules/unspecified-encoding/) (`unspecified-encoding`) does not check for uses of `pathlib.Path.open()` without an explicit encoding argument.
But with Pylint, [W1514](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/unspecified-encoding.html) checks.

I expect Ruff to behave in the same way as Pylint.

### sample.py

```python
from pathlib import Path

with open("foo.txt") as f:
    f.readline()


with Path("foo.txt").open() as f:
    f.readline()
```

### Checks with Ruff

```
$ ruff --version
ruff 0.4.2

$ ruff check --select PLW1514 --preview --output-format=concise --isolated
sample.py:3:6: PLW1514 `open` in text mode without explicit `encoding` argument
Found 1 error.
No fixes available (1 hidden fix can be enabled with the `--unsafe-fixes` option).
```

### Checks with Pylint

```
$ pylint --version
pylint 3.1.0
astroid 3.1.0
Python 3.12.1 (main, Feb  7 2024, 10:25:21) [GCC 11.4.0]

$ pylint sample.py --enable W1514 --disable C
************* Module sample
sample.py:3:5: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)
sample.py:7:5: W1514: Using open without explicitly specifying an encoding (unspecified-encoding)

------------------------------------------------------------------
Your code has been rated at 6.00/10 (previous run: 4.00/10, +2.00)
```

--- a/crates/ruff_linter/resources/test/fixtures/pylint/unspecified_encoding.py
+++ b/crates/ruff_linter/resources/test/fixtures/pylint/unspecified_encoding.py
@@ -70,16 +70,20 @@ open(
     # comment
 )
 
-# Pathlib
+# pathlib
 from pathlib import Path
 
 # Errors.
 Path("foo.txt").open()
 Path("foo.txt").open("w")
+text = Path("foo.txt").read_text()
+Path("foo.txt").write_text(text)
 
 # Non-errors.
 Path("foo.txt").open(encoding="utf-8")
 Path("foo.txt").open("wb")
+text = Path("foo.txt").read_text(encoding="utf-8")
+Path("foo.txt").write_text(text, encoding="utf-8")
 
 # Violation but not detectable
 x = Path("foo.txt")

--- a/crates/ruff_linter/resources/test/fixtures/pylint/unspecified_encoding.py
+++ b/crates/ruff_linter/resources/test/fixtures/pylint/unspecified_encoding.py
@@ -82,8 +82,14 @@ Path("foo.txt").write_text(text)
 # Non-errors.
 Path("foo.txt").open(encoding="utf-8")
 Path("foo.txt").open("wb")
+Path("foo.txt").open(*args)
+Path("foo.txt").open(**kwargs)
 text = Path("foo.txt").read_text(encoding="utf-8")
+text = Path("foo.txt").read_text(*args)
+text = Path("foo.txt").read_text(**kwargs)
 Path("foo.txt").write_text(text, encoding="utf-8")
+Path("foo.txt").write_text(text, *args)
+Path("foo.txt").write_text(text, **kwargs)
 
 # Violation but not detectable
 x = Path("foo.txt")

--- a/crates/ruff_linter/resources/test/fixtures/pylint/unspecified_encoding.py
+++ b/crates/ruff_linter/resources/test/fixtures/pylint/unspecified_encoding.py
@@ -69,3 +69,18 @@ open(
     (("test.txt")),
     # comment
 )
+
+# Pathlib
+from pathlib import Path
+
+# Errors.
+Path("foo.txt").open()
+Path("foo.txt").open("w")
+
+# Non-errors.
+Path("foo.txt").open(encoding="utf-8")
+Path("foo.txt").open("wb")
+
+# Violation but not detectable
+x = Path("foo.txt")
+x.open()

--- a/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
@@ -1,11 +1,11 @@
-use anyhow::Result;
 use std::fmt::{Display, Formatter};
 
-use ast::StringLiteralFlags;
+use anyhow::Result;
+
 use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Fix};
 use ruff_macros::{derive_message_formats, violation};
-use ruff_python_ast as ast;
-use ruff_python_ast::Expr;
+use ruff_python_ast::name::QualifiedName;
+use ruff_python_ast::{self as ast, Expr, StringLiteralFlags};
 use ruff_python_semantic::SemanticModel;
 use ruff_text_size::{Ranged, TextRange};
 
@@ -44,7 +44,7 @@ use crate::settings::types::PythonVersion;
 #[violation]
 pub struct UnspecifiedEncoding {
     function_name: String,
-    mode: Mode,
+    mode: ModeArgument,
 }
 
 impl AlwaysFixableViolation for UnspecifiedEncoding {
@@ -56,10 +56,10 @@ impl AlwaysFixableViolation for UnspecifiedEncoding {
         } = self;
 
         match mode {
-            Mode::Supported => {
+            ModeArgument::Supported => {
                 format!("`{function_name}` in text mode without explicit `encoding` argument")
             }
-            Mode::Unsupported => {
+            ModeArgument::Unsupported => {
                 format!("`{function_name}` without explicit `encoding` argument")
             }
         }
@@ -72,9 +72,10 @@ impl AlwaysFixableViolation for UnspecifiedEncoding {
 
 /// PLW1514
 pub(crate) fn unspecified_encoding(checker: &mut Checker, call: &ast::ExprCall) {
-    let Some((function_name, mode)) = Segments::try_from_call(call, checker.semantic())
-        .filter(|segments| is_violation(call, segments))
-        .map(|segments| (segments.to_string(), Mode::from(&segments)))
+    let Some((function_name, mode)) =
+        QualifiedNameWrapper::try_from_call_expression(call, checker.semantic())
+            .filter(|segments| is_violation(call, segments))
+            .map(|segments| (segments.to_string(), segments.mode_argument()))
     else {
         return;
     };
@@ -96,16 +97,22 @@ pub(crate) fn unspecified_encoding(checker: &mut Checker, call: &ast::ExprCall) 
     checker.diagnostics.push(diagnostic);
 }
 
-enum Segments<'a> {
-    Regular(Vec<&'a str>),
+/// Wrapper around [`QualifiedName`] to consider attribute value on `pathlib.Path(...)` calls.
+enum QualifiedNameWrapper<'a> {
+    /// Regular fully-qualified symbol name.
+    Regular(QualifiedName<'a>),
+    /// Attribute value for the `pathlib.Path(...)` call e.g., `open` in
+    /// `pathlib.Path(...).open(...)`.
     Pathlib(&'a str),
 }
 
-impl<'a> Segments<'a> {
-    fn try_from_call(call: &'a ast::ExprCall, semantic: &'a SemanticModel) -> Option<Self> {
+impl<'a> QualifiedNameWrapper<'a> {
+    fn try_from_call_expression(
+        call: &'a ast::ExprCall,
+        semantic: &'a SemanticModel,
+    ) -> Option<Self> {
         if let Some(qualified_name) = semantic.resolve_qualified_name(&call.func) {
-            let segments = qualified_name.segments().to_vec();
-            return Some(Segments::Regular(segments));
+            return Some(QualifiedNameWrapper::Regular(qualified_name));
         }
         // Check for pathlib.Path(...).open(...) or equivalent
         else if let Expr::Attribute(ast::ExprAttribute { attr, value, .. }) = call.func.as_ref() {
@@ -116,25 +123,40 @@ impl<'a> Segments<'a> {
                         matches!(qualified_name.segments(), ["pathlib", "Path"])
                     })
                 {
-                    return Some(Segments::Pathlib(attr));
+                    return Some(QualifiedNameWrapper::Pathlib(attr));
                 }
             }
         }
         None
     }
+
+    fn mode_argument(&self) -> ModeArgument {
+        match self {
+            QualifiedNameWrapper::Regular(qualified_name) => match qualified_name.segments() {
+                ["" | "codecs" | "_io", "open"] => ModeArgument::Supported,
+                ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
+                    ModeArgument::Supported
+                }
+                ["io" | "_io", "TextIOWrapper"] => ModeArgument::Unsupported,
+                _ => ModeArgument::Unsupported,
+            },
+            QualifiedNameWrapper::Pathlib(attr) => match *attr {
+                "open" => ModeArgument::Supported,
+                _ => ModeArgument::Unsupported,
+            },
+        }
+    }
 }
 
-impl Display for Segments<'_> {
+impl Display for QualifiedNameWrapper<'_> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         match self {
-            Segments::Regular(segments) => {
-                if segments[0].is_empty() {
-                    f.write_str(&segments[1..].join("."))
-                } else {
-                    f.write_str(&segments.join("."))
-                }
+            QualifiedNameWrapper::Regular(qualified_name) => {
+                f.write_str(&qualified_name.to_string())
             }
-            Segments::Pathlib(attr) => f.write_str(&format!("pathlib.Path.{attr}")),
+            QualifiedNameWrapper::Pathlib(attr) => {
+                f.write_str(&format!("pathlib.Path(...).{attr}"))
+            }
         }
     }
 }
@@ -188,7 +210,7 @@ fn is_binary_mode(expr: &Expr) -> Option<bool> {
 }
 
 /// Returns `true` if the given call lacks an explicit `encoding`.
-fn is_violation(call: &ast::ExprCall, segments: &Segments) -> bool {
+fn is_violation(call: &ast::ExprCall, segments: &QualifiedNameWrapper) -> bool {
     // If we have something like `*args`, which might contain the encoding argument, abort.
     if call.arguments.args.iter().any(Expr::is_starred_expr) {
         return false;
@@ -203,7 +225,7 @@ fn is_violation(call: &ast::ExprCall, segments: &Segments) -> bool {
         return false;
     }
     match segments {
-        Segments::Regular(segments) => match segments.as_slice() {
+        QualifiedNameWrapper::Regular(qualified_name) => match qualified_name.segments() {
             ["" | "codecs" | "_io", "open"] => {
                 if let Some(mode_arg) = call.arguments.find_argument("mode", 1) {
                     if is_binary_mode(mode_arg).unwrap_or(true) {
@@ -214,8 +236,9 @@ fn is_violation(call: &ast::ExprCall, segments: &Segments) -> bool {
                 // else mode not specified, defaults to text mode
                 call.arguments.find_argument("encoding", 3).is_none()
             }
-            ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
-                let mode_pos = usize::from(segments[1] == "SpooledTemporaryFile");
+            ["tempfile", tempfile_class @ ("TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile")] =>
+            {
+                let mode_pos = usize::from(*tempfile_class == "SpooledTemporaryFile");
                 if let Some(mode_arg) = call.arguments.find_argument("mode", mode_pos) {
                     if is_binary_mode(mode_arg).unwrap_or(true) {
                         // binary mode or unknown mode is no violation
@@ -234,7 +257,7 @@ fn is_violation(call: &ast::ExprCall, segments: &Segments) -> bool {
             }
             _ => false,
         },
-        Segments::Pathlib(attr) => match *attr {
+        QualifiedNameWrapper::Pathlib(attr) => match *attr {
             "open" => {
                 if let Some(mode_arg) = call.arguments.find_argument("mode", 0) {
                     if is_binary_mode(mode_arg).unwrap_or(true) {
@@ -253,28 +276,9 @@ fn is_violation(call: &ast::ExprCall, segments: &Segments) -> bool {
 }
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq)]
-enum Mode {
+enum ModeArgument {
     /// The call supports a `mode` argument.
     Supported,
     /// The call does not support a `mode` argument.
     Unsupported,
-}
-
-impl From<&Segments<'_>> for Mode {
-    fn from(segments: &Segments) -> Self {
-        match segments {
-            Segments::Regular(segments) => match segments.as_slice() {
-                ["" | "codecs" | "_io", "open"] => Mode::Supported,
-                ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
-                    Mode::Supported
-                }
-                ["io" | "_io", "TextIOWrapper"] => Mode::Unsupported,
-                _ => Mode::Unsupported,
-            },
-            Segments::Pathlib(attr) => match *attr {
-                "open" => Mode::Supported,
-                _ => Mode::Unsupported,
-            },
-        }
-    }
 }

--- a/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
@@ -4,8 +4,8 @@ use ast::StringLiteralFlags;
 use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Fix};
 use ruff_macros::{derive_message_formats, violation};
 use ruff_python_ast as ast;
-use ruff_python_ast::name::QualifiedName;
 use ruff_python_ast::Expr;
+use ruff_python_semantic::SemanticModel;
 use ruff_text_size::{Ranged, TextRange};
 
 use crate::checkers::ast::Checker;
@@ -71,13 +71,11 @@ impl AlwaysFixableViolation for UnspecifiedEncoding {
 
 /// PLW1514
 pub(crate) fn unspecified_encoding(checker: &mut Checker, call: &ast::ExprCall) {
-    let Some((function_name, mode)) = checker
-        .semantic()
-        .resolve_qualified_name(&call.func)
-        .filter(|qualified_name| is_violation(call, qualified_name))
-        .map(|qualified_name| (qualified_name.to_string(), Mode::from(&qualified_name)))
+    let Some((function_name, mode)) = Segments::try_from_call(call, checker.semantic())
+        .filter(|segments| is_violation(call, segments))
+        .map(|segments| (segments.to_string(), Mode::from(&segments)))
     else {
-        return check_pathlib(checker, call);
+        return;
     };
 
     let mut diagnostic = Diagnostic::new(
@@ -97,74 +95,45 @@ pub(crate) fn unspecified_encoding(checker: &mut Checker, call: &ast::ExprCall) 
     checker.diagnostics.push(diagnostic);
 }
 
-fn check_pathlib(checker: &mut Checker, call: &ast::ExprCall) {
-    let Expr::Attribute(ast::ExprAttribute { attr, value, .. }) = call.func.as_ref() else {
-        return;
-    };
+enum Segments<'a> {
+    Regular(Vec<&'a str>),
+    Pathlib(&'a str),
+}
 
-    let Expr::Call(ast::ExprCall { func, .. }) = value.as_ref() else {
-        return;
-    };
-
-    if !checker
-        .semantic()
-        .resolve_qualified_name(func)
-        .is_some_and(|qualified_name| matches!(qualified_name.segments(), ["pathlib", "Path"]))
-    {
-        return;
-    }
-
-    match attr.as_str() {
-        "open" => {
-            if let Some(mode_arg) = call.arguments.find_argument("mode", 0) {
-                if is_binary_mode(mode_arg).unwrap_or(true) {
-                    // binary mode or unknown mode is no violation
-                    return;
+impl<'a> Segments<'a> {
+    fn try_from_call(call: &'a ast::ExprCall, semantic: &'a SemanticModel) -> Option<Self> {
+        if let Some(qualified_name) = semantic.resolve_qualified_name(&call.func) {
+            let segments = qualified_name.segments().to_vec();
+            return Some(Segments::Regular(segments));
+        }
+        // Check for pathlib.Path(...).open(...) or equivalent
+        else if let Expr::Attribute(ast::ExprAttribute { attr, value, .. }) = call.func.as_ref() {
+            if let Expr::Call(ast::ExprCall { func, .. }) = value.as_ref() {
+                if semantic
+                    .resolve_qualified_name(func)
+                    .is_some_and(|qualified_name| {
+                        matches!(qualified_name.segments(), ["pathlib", "Path"])
+                    })
+                {
+                    return Some(Segments::Pathlib(attr));
                 }
             }
-
-            if call.arguments.find_argument("encoding", 2).is_some() {
-                // encoding specified so no violation
-                return;
-            }
         }
-        "read_text" => {
-            if call.arguments.find_argument("encoding", 0).is_some() {
-                // encoding specified so no violation
-                return;
-            }
-        }
-        "write_text" => {
-            if call.arguments.find_argument("encoding", 1).is_some() {
-                // encoding specified so no violation
-                return;
-            }
-        }
-        _ => return,
+        None
     }
 
-    let function_name = format!("pathlib.Path.{attr}");
-    let mode = if attr == "open" {
-        Mode::Supported
-    } else {
-        Mode::Unsupported
-    };
-
-    let mut diagnostic = Diagnostic::new(
-        UnspecifiedEncoding {
-            function_name,
-            mode,
-        },
-        call.func.range(),
-    );
-
-    if checker.settings.target_version >= PythonVersion::Py310 {
-        diagnostic.set_fix(generate_keyword_fix(checker, call));
-    } else {
-        diagnostic.try_set_fix(|| generate_import_fix(checker, call));
+    fn to_string(&self) -> String {
+        match self {
+            Segments::Regular(segments) => {
+                if segments[0] == "" {
+                    segments[1..].join(".")
+                } else {
+                    segments.join(".")
+                }
+            }
+            Segments::Pathlib(attr) => format!("pathlib.Path.{attr}"),
+        }
     }
-
-    checker.diagnostics.push(diagnostic);
 }
 
 /// Generate an [`Edit`] for Python 3.10 and later.
@@ -216,7 +185,7 @@ fn is_binary_mode(expr: &Expr) -> Option<bool> {
 }
 
 /// Returns `true` if the given call lacks an explicit `encoding`.
-fn is_violation(call: &ast::ExprCall, qualified_name: &QualifiedName) -> bool {
+fn is_violation(call: &ast::ExprCall, segments: &Segments) -> bool {
     // If we have something like `*args`, which might contain the encoding argument, abort.
     if call.arguments.args.iter().any(Expr::is_starred_expr) {
         return false;
@@ -230,34 +199,53 @@ fn is_violation(call: &ast::ExprCall, qualified_name: &QualifiedName) -> bool {
     {
         return false;
     }
-    match qualified_name.segments() {
-        ["" | "codecs" | "_io", "open"] => {
-            if let Some(mode_arg) = call.arguments.find_argument("mode", 1) {
-                if is_binary_mode(mode_arg).unwrap_or(true) {
-                    // binary mode or unknown mode is no violation
+    match segments {
+        Segments::Regular(segments) => match segments.as_slice() {
+            ["" | "codecs" | "_io", "open"] => {
+                if let Some(mode_arg) = call.arguments.find_argument("mode", 1) {
+                    if is_binary_mode(mode_arg).unwrap_or(true) {
+                        // binary mode or unknown mode is no violation
+                        return false;
+                    }
+                }
+                // else mode not specified, defaults to text mode
+                call.arguments.find_argument("encoding", 3).is_none()
+            }
+            ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
+                let mode_pos = usize::from(segments[1] == "SpooledTemporaryFile");
+                if let Some(mode_arg) = call.arguments.find_argument("mode", mode_pos) {
+                    if is_binary_mode(mode_arg).unwrap_or(true) {
+                        // binary mode or unknown mode is no violation
+                        return false;
+                    }
+                } else {
+                    // defaults to binary mode
                     return false;
                 }
+                call.arguments
+                    .find_argument("encoding", mode_pos + 2)
+                    .is_none()
             }
-            // else mode not specified, defaults to text mode
-            call.arguments.find_argument("encoding", 3).is_none()
-        }
-        ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
-            let mode_pos = usize::from(qualified_name.segments()[1] == "SpooledTemporaryFile");
-            if let Some(mode_arg) = call.arguments.find_argument("mode", mode_pos) {
-                if is_binary_mode(mode_arg).unwrap_or(true) {
-                    // binary mode or unknown mode is no violation
-                    return false;
+            ["io" | "_io", "TextIOWrapper"] => {
+                call.arguments.find_argument("encoding", 1).is_none()
+            }
+            _ => false,
+        },
+        Segments::Pathlib(attr) => match *attr {
+            "open" => {
+                if let Some(mode_arg) = call.arguments.find_argument("mode", 0) {
+                    if is_binary_mode(mode_arg).unwrap_or(true) {
+                        // binary mode or unknown mode is no violation
+                        return false;
+                    }
                 }
-            } else {
-                // defaults to binary mode
-                return false;
+                // else mode not specified, defaults to text mode
+                call.arguments.find_argument("encoding", 2).is_none()
             }
-            call.arguments
-                .find_argument("encoding", mode_pos + 2)
-                .is_none()
-        }
-        ["io" | "_io", "TextIOWrapper"] => call.arguments.find_argument("encoding", 1).is_none(),
-        _ => false,
+            "read_text" => call.arguments.find_argument("encoding", 0).is_none(),
+            "write_text" => call.arguments.find_argument("encoding", 1).is_none(),
+            _ => false,
+        },
     }
 }
 
@@ -269,15 +257,21 @@ enum Mode {
     Unsupported,
 }
 
-impl From<&QualifiedName<'_>> for Mode {
-    fn from(value: &QualifiedName<'_>) -> Self {
-        match value.segments() {
-            ["" | "codecs" | "_io", "open"] => Mode::Supported,
-            ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
-                Mode::Supported
-            }
-            ["io" | "_io", "TextIOWrapper"] => Mode::Unsupported,
-            _ => Mode::Unsupported,
+impl From<&Segments<'_>> for Mode {
+    fn from(segments: &Segments) -> Self {
+        match segments {
+            Segments::Regular(segments) => match segments.as_slice() {
+                ["" | "codecs" | "_io", "open"] => Mode::Supported,
+                ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
+                    Mode::Supported
+                }
+                ["io" | "_io", "TextIOWrapper"] => Mode::Unsupported,
+                _ => Mode::Unsupported,
+            },
+            Segments::Pathlib(attr) => match *attr {
+                "open" => Mode::Supported,
+                _ => Mode::Unsupported,
+            },
         }
     }
 }

--- a/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
@@ -72,10 +72,9 @@ impl AlwaysFixableViolation for UnspecifiedEncoding {
 
 /// PLW1514
 pub(crate) fn unspecified_encoding(checker: &mut Checker, call: &ast::ExprCall) {
-    let Some((function_name, mode)) =
-        QualifiedNameWrapper::try_from_call_expression(call, checker.semantic())
-            .filter(|segments| is_violation(call, segments))
-            .map(|segments| (segments.to_string(), segments.mode_argument()))
+    let Some((function_name, mode)) = Callee::try_from_call_expression(call, checker.semantic())
+        .filter(|segments| is_violation(call, segments))
+        .map(|segments| (segments.to_string(), segments.mode_argument()))
     else {
         return;
     };
@@ -97,25 +96,22 @@ pub(crate) fn unspecified_encoding(checker: &mut Checker, call: &ast::ExprCall) 
     checker.diagnostics.push(diagnostic);
 }
 
-/// Wrapper around [`QualifiedName`] to consider attribute value on `pathlib.Path(...)` calls.
-enum QualifiedNameWrapper<'a> {
-    /// Regular fully-qualified symbol name.
-    Regular(QualifiedName<'a>),
+/// Represents the path of the function or method being called.
+enum Callee<'a> {
+    /// Fully-qualified symbol name of the callee.
+    Qualified(QualifiedName<'a>),
     /// Attribute value for the `pathlib.Path(...)` call e.g., `open` in
     /// `pathlib.Path(...).open(...)`.
     Pathlib(&'a str),
 }
 
-impl<'a> QualifiedNameWrapper<'a> {
+impl<'a> Callee<'a> {
     fn try_from_call_expression(
         call: &'a ast::ExprCall,
         semantic: &'a SemanticModel,
     ) -> Option<Self> {
-        if let Some(qualified_name) = semantic.resolve_qualified_name(&call.func) {
-            return Some(QualifiedNameWrapper::Regular(qualified_name));
-        }
-        // Check for pathlib.Path(...).open(...) or equivalent
-        else if let Expr::Attribute(ast::ExprAttribute { attr, value, .. }) = call.func.as_ref() {
+        if let Expr::Attribute(ast::ExprAttribute { attr, value, .. }) = call.func.as_ref() {
+            // Check for `pathlib.Path(...).open(...)` or equivalent
             if let Expr::Call(ast::ExprCall { func, .. }) = value.as_ref() {
                 if semantic
                     .resolve_qualified_name(func)
@@ -123,16 +119,21 @@ impl<'a> QualifiedNameWrapper<'a> {
                         matches!(qualified_name.segments(), ["pathlib", "Path"])
                     })
                 {
-                    return Some(QualifiedNameWrapper::Pathlib(attr));
+                    return Some(Callee::Pathlib(attr));
                 }
             }
         }
+
+        if let Some(qualified_name) = semantic.resolve_qualified_name(&call.func) {
+            return Some(Callee::Qualified(qualified_name));
+        }
+
         None
     }
 
     fn mode_argument(&self) -> ModeArgument {
         match self {
-            QualifiedNameWrapper::Regular(qualified_name) => match qualified_name.segments() {
+            Callee::Qualified(qualified_name) => match qualified_name.segments() {
                 ["" | "codecs" | "_io", "open"] => ModeArgument::Supported,
                 ["tempfile", "TemporaryFile" | "NamedTemporaryFile" | "SpooledTemporaryFile"] => {
                     ModeArgument::Supported
@@ -140,7 +141,7 @@ impl<'a> QualifiedNameWrapper<'a> {
                 ["io" | "_io", "TextIOWrapper"] => ModeArgument::Unsupported,
                 _ => ModeArgument::Unsupported,
             },
-            QualifiedNameWrapper::Pathlib(attr) => match *attr {
+            Callee::Pathlib(attr) => match *attr {
                 "open" => ModeArgument::Supported,
                 _ => ModeArgument::Unsupported,
             },
@@ -148,15 +149,11 @@ impl<'a> QualifiedNameWrapper<'a> {
     }
 }
 
-impl Display for QualifiedNameWrapper<'_> {
+impl Display for Callee<'_> {
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         match self {
-            QualifiedNameWrapper::Regular(qualified_name) => {
-                f.write_str(&qualified_name.to_string())
-            }
-            QualifiedNameWrapper::Pathlib(attr) => {
-                f.write_str(&format!("pathlib.Path(...).{attr}"))
-            }
+            Callee::Qualified(qualified_name) => f.write_str(&qualified_name.to_string()),
+            Callee::Pathlib(attr) => f.write_str(&format!("pathlib.Path(...).{attr}")),
         }
     }
 }
@@ -210,7 +207,7 @@ fn is_binary_mode(expr: &Expr) -> Option<bool> {
 }
 
 /// Returns `true` if the given call lacks an explicit `encoding`.
-fn is_violation(call: &ast::ExprCall, qualified_name: &QualifiedNameWrapper) -> bool {
+fn is_violation(call: &ast::ExprCall, qualified_name: &Callee) -> bool {
     // If we have something like `*args`, which might contain the encoding argument, abort.
     if call.arguments.args.iter().any(Expr::is_starred_expr) {
         return false;
@@ -225,7 +222,7 @@ fn is_violation(call: &ast::ExprCall, qualified_name: &QualifiedNameWrapper) -> 
         return false;
     }
     match qualified_name {
-        QualifiedNameWrapper::Regular(qualified_name) => match qualified_name.segments() {
+        Callee::Qualified(qualified_name) => match qualified_name.segments() {
             ["" | "codecs" | "_io", "open"] => {
                 if let Some(mode_arg) = call.arguments.find_argument("mode", 1) {
                     if is_binary_mode(mode_arg).unwrap_or(true) {
@@ -257,7 +254,7 @@ fn is_violation(call: &ast::ExprCall, qualified_name: &QualifiedNameWrapper) -> 
             }
             _ => false,
         },
-        QualifiedNameWrapper::Pathlib(attr) => match *attr {
+        Callee::Pathlib(attr) => match *attr {
             "open" => {
                 if let Some(mode_arg) = call.arguments.find_argument("mode", 0) {
                     if is_binary_mode(mode_arg).unwrap_or(true) {

--- a/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
@@ -77,13 +77,63 @@ pub(crate) fn unspecified_encoding(checker: &mut Checker, call: &ast::ExprCall) 
         .filter(|qualified_name| is_violation(call, qualified_name))
         .map(|qualified_name| (qualified_name.to_string(), Mode::from(&qualified_name)))
     else {
-        return;
+        return check_pathlib(checker, call);
     };
 
     let mut diagnostic = Diagnostic::new(
         UnspecifiedEncoding {
             function_name,
             mode,
+        },
+        call.func.range(),
+    );
+
+    if checker.settings.target_version >= PythonVersion::Py310 {
+        diagnostic.set_fix(generate_keyword_fix(checker, call));
+    } else {
+        diagnostic.try_set_fix(|| generate_import_fix(checker, call));
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+fn check_pathlib(checker: &mut Checker, call: &ast::ExprCall) {
+    let Expr::Attribute(ast::ExprAttribute { attr, value, .. }) = call.func.as_ref() else {
+        return;
+    };
+
+    if attr != "open" {
+        return;
+    }
+
+    let Expr::Call(ast::ExprCall { func, .. }) = value.as_ref() else {
+        return;
+    };
+
+    if !checker
+        .semantic()
+        .resolve_qualified_name(func)
+        .is_some_and(|qualified_name| matches!(qualified_name.segments(), ["pathlib", "Path"]))
+    {
+        return;
+    }
+
+    if let Some(mode_arg) = call.arguments.find_argument("mode", 0) {
+        if is_binary_mode(mode_arg).unwrap_or(true) {
+            // binary mode or unknown mode is no violation
+            return;
+        }
+    }
+
+    if call.arguments.find_argument("encoding", 2).is_some() {
+        // encoding specified so no violation
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(
+        UnspecifiedEncoding {
+            function_name: "pathlib.Path.open".to_string(),
+            mode: Mode::Supported,
         },
         call.func.range(),
     );

--- a/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
@@ -1,4 +1,5 @@
 use anyhow::Result;
+use std::fmt::{Display, Formatter};
 
 use ast::StringLiteralFlags;
 use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Fix};
@@ -121,17 +122,19 @@ impl<'a> Segments<'a> {
         }
         None
     }
+}
 
-    fn to_string(&self) -> String {
+impl Display for Segments<'_> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         match self {
             Segments::Regular(segments) => {
-                if segments[0] == "" {
-                    segments[1..].join(".")
+                if segments[0].is_empty() {
+                    f.write_str(&segments[1..].join("."))
                 } else {
-                    segments.join(".")
+                    f.write_str(&segments.join("."))
                 }
             }
-            Segments::Pathlib(attr) => format!("pathlib.Path.{attr}"),
+            Segments::Pathlib(attr) => f.write_str(&format!("pathlib.Path.{attr}")),
         }
     }
 }

--- a/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
@@ -102,10 +102,6 @@ fn check_pathlib(checker: &mut Checker, call: &ast::ExprCall) {
         return;
     };
 
-    if attr != "open" {
-        return;
-    }
-
     let Expr::Call(ast::ExprCall { func, .. }) = value.as_ref() else {
         return;
     };
@@ -118,22 +114,46 @@ fn check_pathlib(checker: &mut Checker, call: &ast::ExprCall) {
         return;
     }
 
-    if let Some(mode_arg) = call.arguments.find_argument("mode", 0) {
-        if is_binary_mode(mode_arg).unwrap_or(true) {
-            // binary mode or unknown mode is no violation
-            return;
+    match attr.as_str() {
+        "open" => {
+            if let Some(mode_arg) = call.arguments.find_argument("mode", 0) {
+                if is_binary_mode(mode_arg).unwrap_or(true) {
+                    // binary mode or unknown mode is no violation
+                    return;
+                }
+            }
+
+            if call.arguments.find_argument("encoding", 2).is_some() {
+                // encoding specified so no violation
+                return;
+            }
         }
+        "read_text" => {
+            if call.arguments.find_argument("encoding", 0).is_some() {
+                // encoding specified so no violation
+                return;
+            }
+        }
+        "write_text" => {
+            if call.arguments.find_argument("encoding", 1).is_some() {
+                // encoding specified so no violation
+                return;
+            }
+        }
+        _ => return,
     }
 
-    if call.arguments.find_argument("encoding", 2).is_some() {
-        // encoding specified so no violation
-        return;
-    }
+    let function_name = format!("pathlib.Path.{attr}");
+    let mode = if attr == "open" {
+        Mode::Supported
+    } else {
+        Mode::Unsupported
+    };
 
     let mut diagnostic = Diagnostic::new(
         UnspecifiedEncoding {
-            function_name: "pathlib.Path.open".to_string(),
-            mode: Mode::Supported,
+            function_name,
+            mode,
         },
         call.func.range(),
     );

--- a/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
+++ b/crates/ruff_linter/src/rules/pylint/rules/unspecified_encoding.rs
@@ -210,7 +210,7 @@ fn is_binary_mode(expr: &Expr) -> Option<bool> {
 }
 
 /// Returns `true` if the given call lacks an explicit `encoding`.
-fn is_violation(call: &ast::ExprCall, segments: &QualifiedNameWrapper) -> bool {
+fn is_violation(call: &ast::ExprCall, qualified_name: &QualifiedNameWrapper) -> bool {
     // If we have something like `*args`, which might contain the encoding argument, abort.
     if call.arguments.args.iter().any(Expr::is_starred_expr) {
         return false;
@@ -224,7 +224,7 @@ fn is_violation(call: &ast::ExprCall, segments: &QualifiedNameWrapper) -> bool {
     {
         return false;
     }
-    match segments {
+    match qualified_name {
         QualifiedNameWrapper::Regular(qualified_name) => match qualified_name.segments() {
             ["" | "codecs" | "_io", "open"] => {
                 if let Some(mode_arg) = call.arguments.find_argument("mode", 1) {

